lib/smf/smf.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2021 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "smf.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(smf);
+
+/*
+ * Private structure (to this file) used to track state machine context.
+ * The structure is not used directly, but instead to cast the "internal"
+ * member of the smf_ctx structure.
+ */
+struct internal_ctx {
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
+	bool new_state : 1;
+#endif
+	bool terminate : 1;
+	bool exit      : 1;
+};
+
+void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
+{
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
+	const struct smf_state *tmp_state[CONFIG_SMF_ANCESTOR_NUM];
+#endif
+	struct internal_ctx * const internal = (void *) ctx->internal;
+
+	internal->exit = false;
+	internal->terminate = false;
+	ctx->current = init_state;
+	ctx->previous = NULL;
+	ctx->terminate_val = 0;
+
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
```

Comment:
I suggest breaking this block out as a separate function. Something like ``static void smf_set_initial_ancestor(struct smf_ctx *ctx, const struct smf_state *init_state);``

If you do this and add the helper Kconfig symbol, then you can avoid the #ifdefs completely.

Reply:
Done